### DISCUSSION TITLE
Issue when using middleware, BackgroundTasks and request.json()

### DISCUSSION THREAD

**Discussion by tahayk** [fastapi/fastapi#9381] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import json
from fastapi import FastAPI, Request, BackgroundTasks

async def process_payload(request: Request):
    body: json = None
    if request.method in ["POST", "PUT", "DELETE"]:
        try:
            body = await request.json()
        except Exception as e:
            print(">>>> exception")


app = FastAPI()

@app.middleware('http')
async def custom_middleware(request: Request, call_next):
    response = await call_next(request)
    return response

@app.delete("/users/{user_id}")
async def remove_user(user_id: str, request: Request, background_tasks: BackgroundTasks):
    background_tasks.add_task(process_payload, request=request)
    return {"data": True}
```

### Description

When running the provided code: 
1. Send a delete request through [Postman](https://www.postman.com), it will work
2. Send another request, it will take forever
3. Cancel the previous request and send a new one, it will work

I first noticed this through the browser, but using Postman for this issue is easier as I can attach the [postman collection](https://github.com/tiangolo/fastapi/files/11202673/fastapi_issue.postman_collection.json.zip) that you can import directly

After further inspection, I found that:

- if I remove the `middleware`, `body = await request.json()` will throw an Exception, and everything will work fine
- if I remove `body = await request.json()`, everything will work fine 
- if I keep both the `middleware` and `body = await request.json()`; the Excepltion is never thrown, and Postman will wait forever

Even weirder thing, it always work fine if I use `curl` (`curl -X DELETE 'http://127.0.0.1:8000/users/123'` 🤯)

### Operating System

Linux, macOS

### FastAPI Version

0.95.0

### Python Version

3.11.2

**Comment by 8thgencore** [comment]
The issue you're experiencing might be related to the fact that you're not consuming the request body in your middleware, which means that the body is still available for reading when your route handler tries to read it.

When a client sends a request with a body, FastAPI reads the body and saves it in memory. If you don't consume the body in your middleware, the body will still be in memory when your route handler tries to read it. This can cause a deadlock, where the server is waiting for the client to send more data, and the client is waiting for the server to respond.

To fix this, you can consume the request body in your middleware by reading it and then discarding it. Here's an example of how you can do this:

```python
@app.middleware('http')
async def custom_middleware(request: Request, call_next):
    body = b""
    async for chunk in request.stream():
        body += chunk
    request._body = body
    response = await call_next(request)
    return response
```

This code reads the request body using , which returns an async iterator that you can iterate over to read the body in chunks. The body is then saved in memory using `request.stream()` `request._body`

By doing this, you're consuming the request body in the middleware, which means that it won't be available for reading when your route handler tries to read it. This should fix the deadlock issue you're experiencing.

**Reply by tahayk** [reply]
Thank you @8thgencore for the explanation, 
In the example above, I didn't provide any body for the request, is that the same case treated by the code you provided ? 

In addition, you said "it won't be available for reading when your route handler tries to read it"; 
for the example above, I made the code simple by integrating the `process_payload` function in the `background_tasks ` of the route handler, 
but my real use case I'm calling `process_payload` inside a custom `route_class` as a background-task for **all** routes, to   check if the body is available, and make some processing according to that.
So in the example above, the route handler didn't need the payload, but for the rest of the routes, some of them will need the payload and some others won't, some of them will have a payload and some others won't, so the fix I'm looking for should keep the payload available for the route without discarding it in the middleware.
Please let me know if my use case it not clear, I can try and make an example (but it won't be a simple one)

**Comment by tahayk** [comment, marked as answer]
@8thgencore  I found a way to make it work, 
I just have to change the router handler from: 
`async def remove_user(user_id: str, request: Request, background_tasks: BackgroundTasks)`
to:
`async def remove_user(user_id: str, request: Request, background_tasks: BackgroundTasks, _=Body(None)):`

Not sure if this is the correct solution for my use case (or the solution in general).

**Comment by tahayk** [comment]
I'm closing this question/issue for now as I'm satisfied with the solution, feel free to open if needed.
